File: nodejs/tests/transport/smbus_test.js
'use strict';

const { SMBusTransport } = require('../../packages/periph/src/transport/smbus');

const I2C_BUS  = parseInt(process.env.I2C_BUS  || '1',  10);
const I2C_ADDR = parseInt(process.env.I2C_ADDR  || '0x40', 16);

let passed = 0;
let failed = 0;

function checkTrue(label, condition) {
    if (condition) { console.log('PASS', label); passed++; }
    else           { console.log('FAIL', label); failed++; }
}

// --- offline: address validation ---

try {
    new SMBusTransport(I2C_BUS, 0x07);
    checkTrue('addr 0x07 rejected', false);
} catch (e) {
    checkTrue('addr 0x07 rejected', e instanceof RangeError);
}

try {
    new SMBusTransport(I2C_BUS, 0x78);
    checkTrue('addr 0x78 rejected', false);
} catch (e) {
    checkTrue('addr 0x78 rejected', e instanceof RangeError);
}

// --- online: basic I/O without PEC ---

const transport = new SMBusTransport(I2C_BUS, I2C_ADDR);

const readBuf = transport.read(1);
checkTrue('read returns 1 byte', readBuf.length === 1);

transport.write(Buffer.from([0x00]));
checkTrue('write accepted', true);

const wrBuf = transport.writeRead(Buffer.from([0x00]), 1);
checkTrue('writeRead returns 1 byte', wrBuf.length === 1);

// --- online: write with PEC enabled ---

const transportPec = new SMBusTransport(I2C_BUS, I2C_ADDR, true);
transportPec.write(Buffer.from([0x00]));
checkTrue('write with PEC accepted', true);
transportPec.close();

transport.close();

console.log(`===DONE: ${passed} passed, ${failed} failed===`);
process.exit(failed === 0 ? 0 : 1);

File: python/tests/transport/smbus_test.py
import sys

from periph.transport.smbus_auto import SMBusTransport

passed = 0
failed = 0


def check_true(label, condition):
    global passed, failed
    if condition:
        print('PASS', label)
        passed += 1
    else:
        print('FAIL', label)
        failed += 1


# --- offline: address validation ---
# SMBus restricts addresses to 0x08–0x77; addresses outside this range must
# be rejected before any bus access.

try:
    SMBusTransport(0x07)
    check_true('addr 0x07 rejected', False)
except ValueError:
    check_true('addr 0x07 rejected', True)

try:
    SMBusTransport(0x78)
    check_true('addr 0x78 rejected', False)
except ValueError:
    check_true('addr 0x78 rejected', True)

# --- online: bus scan ---

found = []
for probe_addr in range(0x08, 0x78):
    t = SMBusTransport(probe_addr)
    try:
        t.read(1)
        found.append(probe_addr)
    except OSError:
        pass
    finally:
        t.close()

check_true('bus scan: at least one device found', len(found) > 0)
print('devices found:', [hex(a) for a in found])

# --- online: PEC ---
# PEC appends a CRC-8 byte to every write. Whether the device under test
# accepts the extra byte is not a transport concern; we verify the transport
# computes and sends the CRC without crashing.

if found:
    t = SMBusTransport(found[0], pec=True)
    try:
        t.write(bytes([0x00]))
    except OSError:
        pass  # device rejected CRC-appended byte; transport layer worked correctly
    finally:
        t.close()
    check_true('PEC write: no transport crash', True)

print('===DONE: {} passed, {} failed==='.format(passed, failed))
try:
    sys.exit(0 if failed == 0 else 1)
except AttributeError:
    pass  # MicroPython has no sys.exit
